### PR TITLE
bug: fix year formatting in spanner::Date

### LINKED ISSUE #1083: Pad years to 4 digits in spanner::Date objects.
When converting a `spanner::Date` to a proto we should pad the year to 4 digits.  Easy enough to fix but we should write a unit test (and probably an integration test too). See [the documentation](https://cloud.google.com/spanner/docs/data-types#date-type) for more details.

### PR DESCRIPTION
**Pull request by coryan**
The year is required to be 0-padded to 4 digits, the month and date
padding seems to be optional.

Introduce a new integration test to verify data types can be written and
read back.

Fixes #1083

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/1085)

### CODE REVIEW

File: google/cloud/spanner/internal/date.cc [hunk 1]
```diff
@@ -25,7 +25,7 @@ namespace internal {
 
 std::string DateToString(Date d) {
   std::array<char, sizeof "-9223372036854775808-01-01"> buf;
-  std::snprintf(buf.data(), buf.size(), "%" PRId64 "-%02d-%02d", d.year(),
+  std::snprintf(buf.data(), buf.size(), "%04" PRId64 "-%02d-%02d", d.year(),
```
Comment (devbww) [inline google/cloud/spanner/internal/date.cc]: Oops.  Thanks.

File: google/cloud/spanner/integration_tests/data_types_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  RowType read_back;
+  auto commit_result = client_->Commit(
+      [&client, &read_back](Transaction const& txn) -> StatusOr<Mutations> {
+        auto result = client.ExecuteDml(
+            txn,
+            SqlStatement(
+                "INSERT INTO DataTypes (Id, DateValue, StringValue) "
+                "VALUES(@id, @date, @event)",
+                {{"id", Value("ReadWriteDate-1")},
+                 {"date", Value(Date(161, 3, 8))},
```
Comment (devbww) [inline google/cloud/spanner/integration_tests/data_types_integration_test.cc]: Perhaps we should add a "> 9999" case to see how they handle their "YYYY: Four-digit year" edict.

File: google/cloud/spanner/internal/date.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/internal/date.cc]: no problem.

File: google/cloud/spanner/integration_tests/data_types_integration_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/spanner/integration_tests/data_types_integration_test.cc]: Anything over 9999 results in an error.  That is consistent with the [documentation](https://cloud.google.com/spanner/docs/data-types#date-type); which says  the valid range is `0001-01-01` to `9999-12-31`

**Pull request merged**